palace/drivers/basesolver.cpp

```diff
@@ -84,6 +89,282 @@ BaseSolver::BaseSolver(const IoData &iodata, bool root, int size, int num_thread
+      const fs::path root_dir(post_dir);
+      for (const auto &f : fs::directory_iterator(root_dir))
+      {
+        if (f.is_regular_file())
+        {
+          fs::copy(f, out_dir / fs::path(f).filename(), options);
+        }
+      }
+      if (fs::exists(root_dir / "paraview"))
+      {
+        fs::copy(root_dir / "paraview", out_dir / "paraview", options);
+      }
+    }
+    Mpi::Barrier(comm);
+  };
+
+  const bool use_amr = param.adapt_max_its > 0;
+  const bool use_coarsening = mesh.back()->Nonconforming() && param.use_coarsening;
+  if (use_amr && mesh.size() > 1)
+  {
+    Mpi::Print("{}\n", "Flattening mesh sequence: AMR will start from the final mesh in "
+                       "the refinement sequence.");
+    mesh.erase(mesh.begin(), mesh.end() - 1);
+    constexpr bool refine = true, fix_orientation = true;
+    mesh.back()->Finalize(refine, fix_orientation);
+  }
+
+  int iter = 0;
+  auto indicators_and_ntdof = Solve(mesh);
+  const auto &indicators = indicators_and_ntdof.first;
+  const auto &ntdof = indicators_and_ntdof.second;
+  if (use_amr)
+  {
+    const auto is_transient = dynamic_cast<const TransientSolver *>(this) != nullptr;
+    if (is_transient)
+    {
+      MFEM_ABORT("AMR is not currently supported for transient simulations!");
+    }
+    else
+    {
+      Mpi::Print("\nAdaptive Mesh Refinement Parameters:\n");
+      Mpi::Print("MaxIter: {}, Tolerance: {:.3e}\n\n", param.adapt_max_its,
+                 param.adapt_tolerance);
+      SavePostProcess(iter);  // Save the initial solution
+    }
+  }
+
+  // Collection of all tests that might exhaust resources.
+  auto exhausted_resources = [&]()
+  {
+    bool ret = false;
+    // Run out of DOFs, and coarsening isn't allowed.
+    ret |= (!use_coarsening && ntdof > param.dof_limit);
+    // Run out of iterations.
+    ret |= iter >= param.adapt_max_its;
+    return ret;
+  };
+
+  while (indicators.Norml2(comm) > param.adapt_tolerance && !exhausted_resources())
+  {
+    BlockTimer bt_adapt(Timer::ADAPTATION);
+    Mpi::Print("Adaptation iteration {}: Initial Error Indicator: {:.3e}, DOF: {}, DOF "
+               "Limit: {}\n",
+               ++iter, indicators.Norml2(comm), ntdof, param.dof_limit);
+    if (ntdof < param.dof_limit)
+    {
+      // Mark.
+      const auto threshold =
+          utils::ComputeDorflerThreshold(param.update_fraction, indicators.Local());
+      const auto marked_elements = MarkedElements(threshold, indicators.Local());
+
+      // Refine.
+      const auto initial_elem_count = mesh.back()->GetGlobalNE();
+      mesh.back()->GeneralRefinement(marked_elements, -1, param.max_nc_levels);
+      const auto final_elem_count = mesh.back()->GetGlobalNE();
+      Mpi::Print("Mesh refinement added {} elements. Initial: {}, Final: {}\n",
+                 final_elem_count - initial_elem_count, initial_elem_count,
+                 final_elem_count);
+    }
+    else if (use_coarsening)
```

Comment:
Feel free to correct me if this is a dumb question or thought here, but it seems like coarsening should be more useful than only after the final solve when `ntdof >= dof_limit`. Every time we solve, we get a new estimate which could in principle expose elements for coarsening, right? Now, maybe we only want to try to coarsen after a few rounds of refinement, for which we could have a config file option, but I don't see the value in coarsening only after you've hit your limit of the maximum size.